### ISSUE TITLE
Question on assigning pre-defined numbers of CUs to run on different pilots

### ISSUE THREAD

**Issue by mingtaiha** [radical-cybertools/radical.pilot#1315]
Hello!

I am trying to run an experiment where a pre-defined number of CUs to execute on pilots on Gordon, SuperMIC and Stampede (e.g. I want to run 5 CUs on Gordon, 8 CUs on SuperMIC, and 7 CUs on Stampede) and use `radical.analytics` to look at the output. I get errors when I run my script (detailed below), so I would like any input on whether I am using RP correctly to run my experiment.

Here's the script
https://gist.github.com/mingtaiha/6c46615952c53c51898bf752bfca14eb

To overcome the unit scheduling, I created a pmgr and umgr for each pilot starting at [L33](https://gist.github.com/mingtaiha/6c46615952c53c51898bf752bfca14eb#file-baseline_conf-py-L33), and I create and submit the desired number of CUs to the umgr associated with the pilot starting at [L67](https://gist.github.com/mingtaiha/6c46615952c53c51898bf752bfca14eb#file-baseline_conf-py-L67). Starting at [L84](https://gist.github.com/mingtaiha/6c46615952c53c51898bf752bfca14eb#file-baseline_conf-py-L84) I wait for my for all units on all umgrs to finish. Here, I am assuming that `wait_units()` blocks when there are unfinished units belonging to the given umgr, and returns immediately otherwise.

When I run this script, I get the following errors:
```
2017-04-16 13:25:43,061: radical.saga.cpi    : pmgr.launching.1                : Thread-1       : ERROR   : Exception in job monitoring thread: Insufficient system resources: Insufficien
t system resources: read from process failed '[Errno 5] Input/output error' : ( 
20966.0:DONE:0   
47871.0:RUNNING: 
47871.0:DONE:0   
46728.0:RUNNING: 
46728.0:DONE:0   
20441.0:RUNNING: 
20441.0:DONE:0   
43403.0:RUNNING: 
43403.0:DONE:0   
29161.0:RUNNING: 
29161.0:DONE:0   
Shared connection to supermic.cct-lsu.xsede.org closed.
) (()) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_exceptions.py +64 (translate_exception)  :  e = se.NoSuccess ("Insufficient system resources: %s" % cmsg))
2017-04-16 13:25:43,061: radical.saga.cpi    : pmgr.launching.1                : Thread-1       : ERROR   : Cancel job monitoring for gsissh://supermic.cct-lsu.xsede.org:2222/
``` 

```
2017-04-16 13:25:52,023: radical.saga.pty    : pmgr.launching.2                : Thread-4752    : ERROR   : flushing failed                                                                                 
Traceback (most recent call last):
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py", line 174, in flush
    tmp = self.read (timeout=0.1)
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py", line 679, in read
    % (e, self.tail))
NoSuccess: read from process failed '[Errno 5] Input/output error' : (tstrap_1.err
   StdIn=/dev/null
   StdOut=/work/03662/tg829619/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017271.0002/pilot.0002//bootstrap_1.out
   Power= SICP=0

PROMPT-0->Shared connection to stampede.tacc.utexas.edu closed.
) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py +679 (read)  :  % (e, self.tail)))
```

The pilots on Stampede and SuperMIC finished earlier than pilot on Comet. The problem only occurs when the pmgrs are waiting on those two pilots. I get profiles from Comet, but I don't get any agent profiles from Stampede or SuperMIC. However, the `umgr.000*.prof` shows all of the `AGENT` and `UMGR` states. The problem seems to be when the pmgrs are waiting for their pilots.

**Comment by andre-merzky** [comment]
Hey Ming, 

the script looks ok to me, really.  You can simplify it somewhat, but it should be correct as it is.  The simplification would be along these lines:

```py
n_pilots = 3
n_units = 10
pmgr = rp.PilotManager(session)
pd = rp.PilotDescription()
pd.resource = 'local.localhost'
pd.cores = 2

pilots = pmgr.submit_pilots([pd] * n_pilots)

umgrs = list()
for pilot in pilots:
  umgr = rp.UnitManager(session)
  umgr.add_pilots(pilot)
  umgrs.append(umgr)

cud = rp.ComputeUnitDescription()
cud.executable = 'true'
for umgr in umgrs:
  umgr.submit_units([cud] * n_units)
for umgr in umgrs:
  umgr.wait_units()
```

Basically, you can start your pilots with a single pilot manager - but indeed you want to use multiple unit managers with one pilot each.

Now, the above has nothing to do with your problem I'm afraid, or at least not directly: you seem to be running against an (gsi)ssh-level problem.  Can you please set

```sh
export RADICAL_SAGA_PTY_VERBOSE=DEBUG
export RADICAL_SAGA_PTY_LOG_TGT=pty.log
```

and run again, and then post the resulting `pty.log`?  That should contain some more information about the ssh layer.

One thing you may want to make sure is that you do not call `session.close()` while still waiting for the pilots - that will close all ssh connections.  But from your excerpt above it looks like the first exception is already from the ssh layer, and it looks like that is happening before the `session.close()`, but please make sure that's the case.

**Comment by mingtaiha** [comment]
Thanks for the input! I re-ran my script without making the changes you suggested. I noticed that my x509 certificate expired during my run. I re-ran my script with a new x509 certificate and the script ended gracefully. I will try running with your suggestions since it's a more efficient way.


On a related note... From my successful run (without your suggestions), I only got the agent profiles from the pilot which last finished. Is there a way for me to get the other agent profiles? Also, is `radical.analytics` capable of looking at profiles with multiple umgrs?

**Comment by andre-merzky** [comment]
Ah, the X509-expire explains things indeed, thanks.

Analytics should be able to handle multiple umgrs.  We did not test this, the events in a session should be independent from the umgr uid.  Please let us know if you hit any snafus there!

You should be able to retrieve profile for all pilots which reached ACTIVE state.  Can you please check if profiles exist in the remote sandboxes?  If so, please again set the`….PTY_VERBOSE` flags, and try to retrieve them again.  Please send the resulting logfiles, and some instructions on how you tried to retrieve them.

Thanks, Andre.

**Comment by mingtaiha** [comment]
I ran the script again using the `VERBOSE` flags. My x509 certificate should still be valid, but I get the following error message again when waiting for the umgr to return:

```
2017-04-17 13:50:56,508: radical.saga.cpi    : pmgr.launching.0                : Thread-4       : ERROR   : Exception in job monitoring thread: Insufficient system resources: Insufficient system resources: read from process failed '[Errno 5] Input/output error' : ( 
46728.0:DONE:0   
20441.0:RUNNING: 
20441.0:DONE:0   
43403.0:RUNNING: 
43403.0:DONE:0   
29161.0:RUNNING: 
29161.0:DONE:0   
47757.0:RUNNING: 
47757.0:DONE:0   
11712.0:RUNNING: 
11712.0:DONE:0   
Shared connection to supermic.cct-lsu.xsede.org closed.
) (()) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_exceptions.py +64 (translate_exception)  :  e = se.NoSuccess ("Insufficient system resources: %s" % cmsg))
2017-04-17 13:50:56,508: radical.saga.cpi    : pmgr.launching.0                : Thread-4       : ERROR   : Cancel job monitoring for gsissh://supermic.cct-lsu.xsede.org:2222/
```

I also get the following error message at the off of
[pty.log.txt](https://github.com/radical-cybertools/radical.pilot/files/925995/pty.log.txt) when the pmgr is trying to close the connection with SuperMIC.

```
2017-04-17 13:50:56,504: radical.saga.pty    : pmgr.launching.0                : Thread-4       : DEBUG   : read : [  220] [   57] (Shared connection to supermic.cct-lsu.xsede.org closed.\n)
2017-04-17 13:50:56,506: radical.saga.pty    : pmgr.launching.0                : Thread-4       : DEBUG   : Traceback (most recent call last):
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py", line 729, in find
    data = self.read (timeout=_POLLDELAY)
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py", line 679, in read
    % (e, self.tail))
NoSuccess: read from process failed '[Errno 5] Input/output error' : ( 
46728.0:DONE:0   
20441.0:RUNNING: 
20441.0:DONE:0   
43403.0:RUNNING: 
43403.0:DONE:0   
29161.0:RUNNING: 
29161.0:DONE:0   
47757.0:RUNNING: 
47757.0:DONE:0   
11712.0:RUNNING: 
11712.0:DONE:0   
Shared connection to supermic.cct-lsu.xsede.org closed.
) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py +679 (read)  :  % (e, self.tail)))

2017-04-17 13:50:56,507: radical.saga.pty    : pmgr.launching.0                : Thread-4       : DEBUG   : Traceback (most recent call last):
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_shell.py", line 400, in find
    return self.pty_shell.find (patterns, timeout=timeout)
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_process.py", line 790, in find
    raise ptye.translate_exception (e, "(%s)" % data)
NoSuccess: Insufficient system resources: read from process failed '[Errno 5] Input/output error' : ( 
46728.0:DONE:0   
20441.0:RUNNING: 
20441.0:DONE:0   
43403.0:RUNNING: 
43403.0:DONE:0   
29161.0:RUNNING: 
29161.0:DONE:0   
47757.0:RUNNING: 
47757.0:DONE:0   
11712.0:RUNNING: 
11712.0:DONE:0   
Shared connection to supermic.cct-lsu.xsede.org closed.
) (()) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/utils/pty_exceptions.py +64 (translate_exception)  :  e = se.NoSuccess ("Insufficient system resources: %s" % cmsg))
```

**Comment by andre-merzky** [comment]
Bugger.  Can you please make the complete set of logfiles available?  thanks!

**Comment by mingtaiha** [comment]
Sorry for the late reply. Here are all the logs from the local machine and agent logs.
[rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017273.0001.zip](https://github.com/radical-cybertools/radical.pilot/files/930792/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017273.0001.zip)

**Comment by andre-merzky** [comment]
I am afraid that `RADICAL_SAGA_PTY_VERBOSE` was not set for this run...  or maybe you missed to include the pty log file, if that was separate?

**Comment by mingtaiha** [comment]
Here's the `pty.log`. Sorry about that
[pty.log.txt](https://github.com/radical-cybertools/radical.pilot/files/937749/pty.log.txt)

**Comment by mingtaiha** [comment]
@andre-merzky Did you manage to reproduce the error? Lmk if I can do anything to help.

**Comment by andre-merzky** [comment]
Yes - and after fixing the first, I found more popping up in this use case.  I am working on a proper implementation of direct binding, which we have been missing out on for a while.  So bad news is that you can't use it yet, good news is that I am on it finally :)

**Comment by andre-merzky** [comment]
Ming, can you please have a look at the `feature/early_binding` branch?  An example using this is [here](https://gist.github.com/andre-merzky/8aeed7c5e1ea3ae0042a2e2a705ea6de).  Note that line 48 is like this:

```
   cud.pilot = pilot_ids[i]
```

which will tell the umgr scheduler that the unit *must* be scheduled to this pilot.  So that mechanism works with a single umgr, pmgr, session.  There is currently one limitation which I have to resolve: the `pmgr.wait_pilots(state=rp.ACTIVE)` call in line 39.  That is not a fundamental problem, but avoids a race condition which needs addressing.

Let me know what you think, and if that syntax makes sense for your use case.

**Comment by mingtaiha** [comment]
Thanks for getting back to me. I think I get the example you gave, as you're just associating a `pilot.uid` to a unit. I get the following error when I try to run the script with the `feature/early_binding` branch :/

```
(ve.exec_model_exp) mingtha@Ubuntu-1604-xenial-64-minimal:~/repos/affinity_model/exp/res_select$ python test_early_binding.py 
Traceback (most recent call last):
  File "test_early_binding.py", line 8, in <module>
    import radical.pilot as rp
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/__init__.py", line 23, in <module>
    from .session                   import Session
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/session.py", line 29, in <module>
    from . import utils         as rpu
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/utils/__init__.py", line 15, in <module>
    from .queue        import *
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/utils/queue.py", line 102, in <module>
    class Queue(ru.Process):
AttributeError: 'module' object has no attribute 'Process'
```

**Comment by andre-merzky** [comment]
Please switch to RU branch 'feature/managed_process'.  Thanks!

...

**Comment by mingtaiha** [comment]
Do I also need to change branch for SAGA? I get the following error
```
Creating and submitting CUs
exception: attribute set is not extensible/private (key pilot) (/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/saga/attributes.py +1068 (_attributes_i_set)  :  raise se.IncorrectState ("attribute set is not extensible/private (key %s)" %  key))
```

**Comment by mingtaiha** [comment]
I tried this again with the newest version of SAGA, as discussed in the hangouts meeting, but I still get the above error. My `radical-stack` is:
```
python            : 2.7.12
virtualenv        : /home/mingtha/ve/ve.exec_model_exp
radical.utils     : v0.45-2-g82050c5@devel
saga-python       : split-16-g4be5f196@devel
radical.pilot     : split-66-ge67c4f3c@devel
radical.analytics : v0.45@devel
```

**Comment by andre-merzky** [comment]
I am afraid the RP commit looks wrong - you should be on 762425c, but you are on e67c4f3c, which is one behind.  Can you please check / update RP?

**Comment by mingtaiha** [comment]
D'oh, I needed to include `--no-deps` when installing! I'm running your example and I'll get back to you on that

**Comment by mingtaiha** [comment]
Hey, I gave the your example a try, running using `localhost`. The execution hangs and I get errors in many of the log files:


```Traceback (most recent call last):                                                                                                                                                                          
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/utils/process.py", line 549, in _run
    if not self.work_cb():
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/utils/pubsub.py", line 231, in work_cb
    _socks = dict(_uninterruptible(self._poll.poll, timeout=1000)) # timeout in ms
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/utils/pubsub.py", line 50, in _uninterruptible
    return f(*args, **kwargs)
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/zmq/sugar/poll.py", line 99, in poll
    return zmq_poll(self.sockets, timeout=timeout)
  File "zmq/backend/cython/_poll.pyx", line 116, in zmq.backend.cython._poll.zmq_poll (zmq/backend/cython/_poll.c:2036)
  File "zmq/backend/cython/checkrc.pxd", line 12, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/_poll.c:2418)
    PyErr_CheckSignals()
KeyboardInterrupt
```

I also get the error in my `pmgr.launching.0.child.log` but I think it is because of the above error.
```
Traceback (most recent call last):
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/pmgr/launching/default.py", line 302, in work
    self._start_pilot_bulk(resource, schema, pilots)
  File "/home/mingtha/ve/ve.exec_model_exp/local/lib/python2.7/site-packages/radical/pilot/pmgr/launching/default.py", line 499, in _start_pilot_bulk
    raise RuntimeError ("SAGA Job state is FAILED.")
RuntimeError: SAGA Job state is FAILED.
```

Here's the log files.
[session_logs.zip](https://github.com/radical-cybertools/radical.pilot/files/971911/session_logs.zip)

**Comment by andre-merzky** [comment]
Oh the joys of Python deployment: it looks like your pilot sandbox doesn't like us anymore.  From the bootstrapper output:

```
cmd: /home/mingtha/radical.pilot.sandbox/ve_localhost/bin/pip install  --src '/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017288.0001/pilot.0000/rp_install/src' --build '/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017288.0001/pilot.0000/rp_install/build' --install-option='--prefix=/home/mingtha/radical.pilot.sandbox/rp.session.Ubuntu-1604-xenial-64-minimal.mingtha.017288.0001/pilot.0000/rp_install' --no-deps radical.utils-v0.45-45-gce49d06-feature-managed-process/
#
Traceback (most recent call last):
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/bin/pip", line 8, in <module>
    load_entry_point('pip==1.4.1', 'console_scripts', 'pip')()
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 318, in load_entry_point
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 2221, in load_entry_point
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 1954, in load
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/pip/__init__.py", line 9, in <module>
    from pip.log import logger
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/pip/log.py", line 7, in <module>
    from pip import backwardcompat
  File "/home/mingtha/radical.pilot.sandbox/ve_localhost/local/lib/python2.7/site-packages/pip/backwardcompat/__init__.py", line 69, in <module>
    from urllib2 import URLError, HTTPError
  File "/usr/lib/python2.7/urllib2.py", line 93, in <module>
    import hashlib
  File "/usr/lib/python2.7/hashlib.py", line 138, in <module>
    _hashlib.openssl_md_meth_names)
AttributeError: 'module' object has no attribute 'openssl_md_meth_names'
```

I would suggest to remove `/home/mingtha/radical.pilot.sandbox/ve_localhost` and to try again, possibly first with a single pilot -- but I am not really sure if that helps, as the underlying problem seems to be on the system Python level...

**Comment by mingtaiha** [comment]
Luckily the new `ve_localhost` does! Deleting the old virtualenv seemed to work. I'll try to it on Supermic and Comet now.

**Comment by andre-merzky** [comment]
Puh, good! :)

**Comment by mingtaiha** [comment]
I successfully tested the changes 16 CUs. I am still testing using more CUs, but the queue on Comet is quite long (1-2 days), so I'm still working on this.

**Comment by andre-merzky** [comment]
Great! The umgr scheduler is independent of pilot locality and number of units.  If it works for 16 CUs on comet, it works - everything would be a separate ticket - so feel free to close...  Thanks!

**Comment by mingtaiha** [comment]
The scheduler seems to check out with 1024 CUs. However, I encounter another error. I'll close this ticket and open a new one to address the issue.
